back-end/src/main/java/index/PsqlCubeSpatialIndexer.java

```diff
@@ -0,0 +1,224 @@
+        bboxStmt.executeUpdate(sql);
+
+        // if this is an empty layer, return
+        if (trans.getDb().equals(""))
+            return ;
+
+        // step 1: set up nashorn environment -- what is nashorn??
+        NashornScriptEngine engine = null;
+        if (! trans.getTransformFunc().equals(""))
+            engine = setupNashorn(trans.getTransformFunc());
+
+        // step 2: looping through query results
+        // what's difference btwn separable and non-separable cases?
+        Statement rawDBStmt = DbConnector.getStmtByDbName(trans.getDb());
+        ResultSet rs = DbConnector.getQueryResultIterator(rawDBStmt, trans.getQuery());
+        int numColumn = rs.getMetaData().getColumnCount();
+        int rowCount = 0;
+        String insertSql = "insert into " + bboxTableName + " values (";
+        // where does the 6 come from?
+        for (int i = 0; i < colNames.size() + 6; i ++) {
+            insertSql += "?, ";
+        }
+        insertSql += "?::cube);";
+        PreparedStatement preparedStmt = dbConn.prepareStatement(insertSql);
+        while (rs.next()) {
+
+            // count log
+            rowCount ++;
+            if (rowCount % 1000000 == 0)
+                System.out.println(rowCount);
+
+            // get raw row
+            ArrayList<String> curRawRow = new ArrayList<>();
+            for (int i = 1; i <= numColumn; i ++)
+                curRawRow.add(rs.getString(i));
+
+            // step 3: run transform function on this tuple
+            ArrayList<String> transformedRow;
+            if (! trans.getTransformFunc().equals(""))
+                transformedRow = getTransformedRow(c, curRawRow, engine);
+            else
+                transformedRow = curRawRow;
+
+            // step 4: calculate bounding boxes
+            ArrayList<Double> curBbox = getBboxCoordinates(l, transformedRow);
+
+            // insert into bbox table
+            for (int i = 0; i < transformedRow.size(); i ++)
+                // preparedStmt.setString(i + 1, transformedRow.get(i).replaceAll("\'", "\'\'"));
+                preparedStmt.setString(i + 1, transformedRow.get(i));
+            for (int i = 0; i < 6; i ++)
+                preparedStmt.setDouble(transformedRow.size() + i + 1, curBbox.get(i));
+
+            double minx, miny, maxx, maxy;
+            minx = curBbox.get(2);
+            miny = curBbox.get(3);
+            maxx = curBbox.get(4);
+            maxy = curBbox.get(5);
+            
+            preparedStmt.setString(transformedRow.size() + 7,
+                    // getPolygonText(minx, miny, maxx, maxy));
+                    getCubeText(minx, miny, maxx, maxy, canvasId));
+            preparedStmt.addBatch();
+
+            if (rowCount % Config.bboxBatchSize == 0) {
+                preparedStmt.executeBatch();
+                DbConnector.commitConnection(Config.databaseName);
```

Comment:
now we use auto commit for postgres, because we do not need transactions for indexing.

Therefore, you don't need to manually commit connections. In other words, `DbConnector.commitConnection` should not be used by any indexers.

Reply:
That's good to know. I removed all of the commitConnection calls so that it's consistent